src/Erl/Data/Map/Internal.purs

```diff
@@ -0,0 +1,175 @@
+module Erl.Data.Map.Internal
+  ( Map
+  , empty
+  , isEmpty
+  , size
+  , insert
+  , singleton
+  , lookup
+  , values
+  , keys
+  , mapWithKey
+  , member
+  , delete
+  , difference
+  , fromFoldable
+  , alter
+  ) where
+
+import Prelude
+
+import Data.Foldable (class Foldable, foldl, foldr)
+import Data.FoldableWithIndex (class FoldableWithIndex, foldlWithIndex)
+import Data.Function.Uncurried (Fn2, mkFn2)
+import Data.List.Lazy as LL
+import Data.Maybe (Maybe(..), maybe, maybe')
+import Data.Traversable (class Traversable, sequenceDefault)
+import Data.Tuple (Tuple(..), uncurry)
+import Data.Unfoldable (class Unfoldable)
+import Erl.Data.List (List)
+
+foreign import data Map :: Type -> Type -> Type
+
+-- | An empty map
+foreign import empty :: forall a b. Map a b
+
+-- | Test if a map is empty
+foreign import isEmpty :: forall a b. Map a b -> Boolean
+
+-- | Calculate the number of key/value pairs in a map
+foreign import size :: forall a b. Map a b -> Int
+
+-- | Insert or replace a key/value pair in a map
+foreign import insert :: forall a b. a -> b -> Map a b -> Map a b
+
+-- | Test if a key is a member of a map
+foreign import member :: forall k a. k -> Map k a -> Boolean
+
+-- | Difference of two maps. Return elements of the first map where
+-- | the keys do not exist in the second map.
+foreign import difference :: forall k a b. Map k a -> Map k b -> Map k a
+
+-- | Delete a key and its corresponding value from a map.
+foreign import delete :: forall k a. k -> Map k a -> Map k a
+
+-- | Get a list of the values contained in a map
+foreign import values :: forall a b. Map a b -> List b
+
+-- | Get a list of the keys contained in a map
+foreign import keys :: forall a b. Map a b -> List a
+
+-- Folds taken from purescript-foreign-object
+foreign import foldMImpl
+  :: forall a b m z.
+  (m -> (z -> m) -> m)
+  -> (z -> a -> b -> m)
+  -> m
+  -> Map a b
+  -> m
+
+foreign import lookupImpl :: forall a b z. z -> (b -> z) -> a -> Map a b -> z
+foreign import mapImpl :: forall k a b. (a -> b) -> Map k a -> Map k b
+foreign import mapWithKeyImpl :: forall k a b. (Fn2 k a b) -> Map k a -> Map k b
+foreign import toUnfoldMImp :: forall f k v. Unfoldable f => Map k v -> f (Tuple k v)
```

Comment:
not sure if this is the correct implementation because of the `f` I assume erlang returns a list?